x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java

```diff
@@ -0,0 +1,315 @@
+        if (fieldName == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires a fieldName");
+        }
+        if (query == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires a " + QUERY_FIELD.getPreferredName() + " value");
+        }
+        this.fieldName = fieldName;
+        this.query = query;
+    }
+
+    public SemanticQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.fieldName = in.readString();
+        this.query = in.readString();
+        this.inferenceResults = in.readOptionalNamedWriteable(InferenceServiceResults.class);
+        if (this.inferenceResults != null) {
+            // The supplier is generally not used after the results are set, but set the supplier to maintain equality when serializing
+            // & deserializing
+            this.inferenceResultsSupplier = new SetOnce<>(this.inferenceResults);
+        }
+    }
+
+    private SemanticQueryBuilder(SemanticQueryBuilder other, SetOnce<InferenceServiceResults> inferenceResultsSupplier) {
+        this.fieldName = other.fieldName;
+        this.query = other.query;
+        this.boost = other.boost;
+        this.queryName = other.queryName;
+        this.inferenceResultsSupplier = inferenceResultsSupplier;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersions.SEMANTIC_TEXT_FIELD_ADDED;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        if (inferenceResultsSupplier != null && inferenceResults == null) {
+            throw new IllegalStateException("Inference results supplier is set, but inference results is null. Missing a rewriteAndFetch?");
+        }
+        out.writeString(fieldName);
+        out.writeString(query);
+        out.writeOptionalNamedWriteable(inferenceResults);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.startObject(fieldName);
+        builder.field(QUERY_FIELD.getPreferredName(), query);
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+        builder.endObject();
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
+        // We cannot fully rewrite the query to a NestedQueryBuilder here because that query builder validates that the path is
+        // registered as a nested object. This is not the case for semantic_text fields; the SemanticTextInferenceResultFieldMapper
+        // metafield mapper indexes inference results for the field using a "shadow" nested field mapping that is coordinated between
+        // this query and the mapper. This "shadow" mapping is not registered with the ES index mappings.
```

Comment:
This is fixed in the branch now so we should be able to use the QueryBuilders directly. 
We also don't need to implement `toQuery` if we're working with the es builders. 
All the logic can live in the rewrite.

Reply:
Understood. It does increase the complexity of the rewrite process, since it will now be:

1. Rewrite on coordinator to get inference results
2. Rewrite on data shard (where you have access to the index/mappers/etc) to create the query builders

There are also some edge cases to work out, such as when we want to return a `MatchNoDocsQuery`. This could happen if the user queries a field that doesn't exist or queries a valid `semantic_text` field that doesn't have any inferenced docs (and thus does not have the subfields required to generate the appropriate sparse/dense vector query).

Also, this work depends on work in your branch. Could we merge this as-is and I can follow up with these changes once your branch is merged?

Reply:
We are using `QueryBuilder`s to build the query now